In [1]:
import numpy as np
import pandas as pd

## Introduction

Q1: Load the dataset.

In [2]:
book_df = pd.read_csv('book_data.csv')
book_df

book_authors                                          book_desc  \
0       Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
1      Madeline Miller  In the house of Helios, god of the sun and mig...   
2          Holly Black  Of course I want to be like them. They’re beau...   
3            A.J. Finn  Anna Fox lives alone—a recluse in her New York...   
4        Tara Westover  An unforgettable memoir in the tradition of Th...   
...                ...                                                ...   
1136    Alyssa Brandon  A teen werewolf finally meets her destined sou...   
1137  Crystal Hana Kim  An emotionally riveting debut novel about war,...   
1138     Leslie  Cohen  This is the classic tale of boy meets girl: Gi...   
1139  Sarah Beth Durst  The battle between vicious spirits and strong-...   
1140      Nisha Sharma  Winnie Mehta was never really convinced that R...   

     book_edition     book_format     book_isbn book_pages  book_rating  \
0             NaN  Kindle Edition           NaN  435 pages         4.33   
1             NaN       Hardcover  9.780000e+12  393 pages         4.34   
2             NaN       Hardcover  9.780000e+12  370 pages         4.18   
3             NaN  Kindle Edition           NaN  449 pages         3.97   
4             NaN       Hardcover  9.780000e+12  334 pages         4.47   
...           ...             ...           ...        ...          ...   
1136          NaN       Hardcover  9.780000e+12  304 pages         3.33   
1137          NaN       Hardcover  9.780000e+12  417 pages         3.69   
1138          NaN  Kindle Edition           NaN  320 pages         3.40   
1139          NaN       Hardcover  9.780000e+12  419 pages         3.86   
1140          NaN       Hardcover  9.780000e+12  304 pages         3.64   

      book_rating_count  book_review_count  \
0                146505              17438   
1                 61357               9369   
2                 62897              11948   
3                175678              20645   
4                128337              16424   
...                 ...                ...   
1136                359                 88   
1137               1255                259   
1138               1300                272   
1139                907                205   
1140               1322                395   

                              book_title  \
0                        The Great Alone   
1                                  Circe   
2                       The Cruel Prince   
3                The Woman in the Window   
4                               Educated   
...                                  ...   
1136                        Bound to You   
1137                     If You Leave Me   
1138  This Love Story Will Self-Destruct   
1139                 The Queen of Sorrow   
1140         My So-Called Bollywood Life   

                                                 genres  \
0     Fiction|Historical|Historical Fiction|Audioboo...   
1     Fantasy|Fiction|Fantasy|Mythology|Historical|H...   
2                       Fantasy|Young Adult|Fairies|Fae   
3     Mystery|Thriller|Fiction|Suspense|Thriller|Mys...   
4             Nonfiction|Autobiography|Memoir|Biography   
...                                                 ...   
1136  Fantasy|Paranormal|Romance|Fantasy|Shapeshifte...   
1137  Historical|Historical Fiction|Fiction|Historic...   
1138  Romance|Fiction|Womens Fiction|Chick Lit|Conte...   
1139                          Fantasy|Young Adult|Adult   
1140  Young Adult|Contemporary|Romance|Fiction|Young...   

                                              image_url  
0     https://images.gr-assets.com/books/1501852423l...  
1     https://images.gr-assets.com/books/1508879575l...  
2     https://images.gr-assets.com/books/1493047153l...  
3     https://images.gr-assets.com/books/1528225499l...  
4     https://images.gr-assets.com/books/1506026635l...  
...                                          

Q2: Find if there are any duplicates book titles. Remove duplicated from the data keeping only the first occurrence of a title. How many unique books are there?

There are duplicate book titles.

In [3]:
book_df[book_df.duplicated('book_title')]

book_authors  \
538                    Mackenzi Lee   
637                     V.E. Schwab   
639                    Sejal Badani   
773               Brandon Sanderson   
794                      John Boyne   
811                Rebecca Traister   
829                     V.E. Schwab   
846                 Victoria Schwab   
857                  Chanel Cleeton   
892                   Marisha Pessl   
945    Janet Evanovich|Lorelei King   
956                     James Comey   
1008                   Sandie Jones   
1124   Bill Clinton|James Patterson   
1126  Douglas Preston|Lincoln Child   
1129                 Mariana Zapata   

                                              book_desc  \
538   \r\nIn this highly anticipated sequel to the N...   
637   Eli Ever and Victor Vale were only medical stu...   
639   From the bestselling author of Trail of Broken...   
773   \r\nFrom Brandon Sanderson, the #1 \r\n\r\nNew...   
794   \r\nThe new novel from the beloved \r\n\r\nNew...   
811   From Rebecca Traister, the New York Times best...   
829   Eli Ever and Victor Vale were only medical stu...   
846   Cassidy Blake's parents are The Inspectres, a ...   
857   Havana, 1958. A sugar baron's daughter, ninete...   
892   Once upon a time, back at Darrow-Harker School...   
945   \r\nStephanie Plum faces the toughest puzzle o...   
956   Listening Length: 9 hours and 5 minutesFormer ...   
1008  A deliciously disturbing, compulsively readabl...   
1124  The President disappears. The world is in shoc...   
1126  What begins as a manhunt for the missing daugh...   
1129  If someone were to ask Jasmine Santos to descr...   

                book_edition     book_format     book_isbn book_pages  \
538                      NaN  Kindle Edition           NaN  464 pages   
637   UK Collector's Edition       Hardcover           NaN  575 pages   
639                      NaN       Paperback  9.780000e+12  412 pages   
773                      NaN  Kindle Edition           NaN  515 pages   
794                      NaN  Kindle Edition           NaN  357 pages   
811                      NaN       Hardcover  9.780000e+12  320 pages   
829                      NaN       Hardcover  9.780000e+12  478 pages   
846                      NaN           ebook  9.780000e+12  272 pages   
857                      NaN       Hardcover  9.780000e+12  596 pages   
892                      NaN       Hardcover           NaN  327 pages   
945                      NaN        Audio CD  9.780000e+12        NaN   
956      Unabridged Download       Audiobook  9.780000e+12   10 pages   
1008                     NaN       Hardcover  9.780000e+12  304 pages   
1124             Large print       Hardcover  9.780000e+12  688 pages   
1126                     NaN  Kindle Edition           NaN  416 pages   
1129             1st edition  Kindle Edition           NaN  524 pages   

      book_rating  book_rating_count  book_review_count  \
538          4.24               6498               1660   
637          4.38               8782               2060   
639          4.45              24372               1799   
773          4.51              10649               1636   
794          4.26               4293                954   
811          4.50               2217                409   
829          4.38               8761               2059   
846          3.97              12358               2793   
857          4.01              16205               2233   
892          3.79               6070               1317   
945          3.86               8714               1032   
956          4.21              26353               4089   
1008         3.85              14060               2726   
1124         3.88              45587               5840   
1126         4.07               9962               1125   
1129         4.35              17515               2121   

                                             book_title  \
538           The Lady's Guide to Petticoats and P

In [4]:
book_df = book_df[~book_df.duplicated('book_title')]
book_df

book_authors                                          book_desc  \
0       Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
1      Madeline Miller  In the house of Helios, god of the sun and mig...   
2          Holly Black  Of course I want to be like them. They’re beau...   
3            A.J. Finn  Anna Fox lives alone—a recluse in her New York...   
4        Tara Westover  An unforgettable memoir in the tradition of Th...   
...                ...                                                ...   
1136    Alyssa Brandon  A teen werewolf finally meets her destined sou...   
1137  Crystal Hana Kim  An emotionally riveting debut novel about war,...   
1138     Leslie  Cohen  This is the classic tale of boy meets girl: Gi...   
1139  Sarah Beth Durst  The battle between vicious spirits and strong-...   
1140      Nisha Sharma  Winnie Mehta was never really convinced that R...   

     book_edition     book_format     book_isbn book_pages  book_rating  \
0             NaN  Kindle Edition           NaN  435 pages         4.33   
1             NaN       Hardcover  9.780000e+12  393 pages         4.34   
2             NaN       Hardcover  9.780000e+12  370 pages         4.18   
3             NaN  Kindle Edition           NaN  449 pages         3.97   
4             NaN       Hardcover  9.780000e+12  334 pages         4.47   
...           ...             ...           ...        ...          ...   
1136          NaN       Hardcover  9.780000e+12  304 pages         3.33   
1137          NaN       Hardcover  9.780000e+12  417 pages         3.69   
1138          NaN  Kindle Edition           NaN  320 pages         3.40   
1139          NaN       Hardcover  9.780000e+12  419 pages         3.86   
1140          NaN       Hardcover  9.780000e+12  304 pages         3.64   

      book_rating_count  book_review_count  \
0                146505              17438   
1                 61357               9369   
2                 62897              11948   
3                175678              20645   
4                128337              16424   
...                 ...                ...   
1136                359                 88   
1137               1255                259   
1138               1300                272   
1139                907                205   
1140               1322                395   

                              book_title  \
0                        The Great Alone   
1                                  Circe   
2                       The Cruel Prince   
3                The Woman in the Window   
4                               Educated   
...                                  ...   
1136                        Bound to You   
1137                     If You Leave Me   
1138  This Love Story Will Self-Destruct   
1139                 The Queen of Sorrow   
1140         My So-Called Bollywood Life   

                                                 genres  \
0     Fiction|Historical|Historical Fiction|Audioboo...   
1     Fantasy|Fiction|Fantasy|Mythology|Historical|H...   
2                       Fantasy|Young Adult|Fairies|Fae   
3     Mystery|Thriller|Fiction|Suspense|Thriller|Mys...   
4             Nonfiction|Autobiography|Memoir|Biography   
...                                                 ...   
1136  Fantasy|Paranormal|Romance|Fantasy|Shapeshifte...   
1137  Historical|Historical Fiction|Fiction|Historic...   
1138  Romance|Fiction|Womens Fiction|Chick Lit|Conte...   
1139                          Fantasy|Young Adult|Adult   
1140  Young Adult|Contemporary|Romance|Fiction|Young...   

                                              image_url  
0     https://images.gr-assets.com/books/1501852423l...  
1     https://images.gr-assets.com/books/1508879575l...  
2     https://images.gr-assets.com/books/1493047153l...  
3     https://images.gr-assets.com/books/1528225499l...  
4     https://images.gr-assets.com/books/1506026635l...  
...                                          

There are 1125 unique books.

Q3: How many books have 2 authors?

In [5]:
book_df[book_df['book_authors'].str.count('\|') == 1]

book_authors  \
8          Greer Hendricks|Sarah Pekkanen   
13                Helen Hoang|Helen Hoang   
19              Amie Kaufman|Jay Kristoff   
44               Audrey Grey|Krystal Wade   
80            Amie Kaufman|Meagan Spooner   
...                                   ...   
1066         Mariko Tamaki|Brooklyn Allen   
1069             Justina Ireland|Annie Wu   
1071            Scott Snyder|Greg Capullo   
1079    Neal Shusterman|Jarrod Shusterman   
1131  Laurie Halse Anderson|Emily Carroll   

                                              book_desc    book_edition  \
8     When you read this book, you will make many as...             NaN   
13    A heartwarming and refreshing debut novel that...  1st US edition   
19    Kady, Ezra, Hanna, and Nik narrowly escaped wi...             NaN   
44    While she sleeps, the whole universe changes.P...     1st Edition   
80    When Earth intercepts a message from a long-ex...             NaN   
...                                                 ...             ...   
1066  Welcome to Miss Qiunzella Thiskwin Penniquiqul...             NaN   
1069  An all-new adventure starring Lando Calrissian...             NaN   
1071  The Dark Multiverse is coming, and nothing wil...             NaN   
1079  When the California drought escalates to catas...   First Edition   
1131  The modern classic Speak is now a graphic nove...             NaN   

         book_format     book_isbn book_pages  book_rating  book_rating_count  \
8          Hardcover  9.780000e+12  346 pages         3.85              89835   
13    Kindle Edition           NaN  336 pages         4.00              33760   
19         Hardcover  9.780000e+12  615 pages         4.58              18423   
44         Paperback  9.780000e+12  316 pages         3.78                347   
80         Hardcover  9.780000e+12  328 pages         3.86               3976   
...              ...           ...        ...          ...                ...   
1066       Hardcover  9.780000e+12  208 pages         4.15                 41   
1069       Hardcover  9.780000e+12  176 pages         3.56                 85   
1071       Hardcover  9.780000e+12  216 pages         3.60               1122   
1079       Hardcover  9.780000e+12  390 pages         4.08               3636   
1131       Hardcover  9.780000e+12  374 pages         4.30               5641   

      book_review_count                 book_title  \
8                 11331        The Wife Between Us   
13                 6502          The Kiss Quotient   
19                 4809                    Obsidio   
44                  182             Our Dark Stars   
80                 1119                  Unearthed   
...                 ...                        ...   
1066                  6  Lumberjanes: The Good Egg   
1069                 25    Star Wars: Lando's Luck   
1071                209         Dark Nights: Metal   
1079                944                        Dry   
1131                908   Speak: The Graphic Novel   

                                                 genres  \
8     Thriller|Fiction|Mystery|Suspense|Thriller|Mys...   
13    Romance|Contemporary|Fiction|Adult|Romance|Con...   
19                          Science Fiction|Young Adult   
44          Science Fiction|Young Adult|Romance|Fantasy   
80                          Science Fiction|Young Adult   
...                                                 ...   
1066  Childrens|Middle Grade|Fantasy|Young Adult|Glb...   
1069  Media Tie In|Star Wars|Science Fiction|Fiction...   
1071  Sequential Art|Comics|Sequential Art|Graphic N...   
1079  Young Adult|Science Fiction|Dystopia|Science F...   
1131  Sequential Art|Graphic Novels|Young Adult|Real...   

                                              image_url  
8     https://images.gr-assets.com/books/1496862122l...  
13    https://images.gr-assets.com/books/1527983633l...  
19    https://images.gr-assets.com/books/1501704611l...  
44    https://images.gr-

Totally 107 books have 2 authors.

## Length

Q4: You will want to create a new column with the integer value of the number of pages. (If you remove rows in this process, please state why.)

I remove the rows whose book_pages is NaN. Then the column can be changed into integer directly.

In [6]:
book_df2 = book_df.dropna(subset = ['book_pages'])
book_df2['book_pages_count'] = book_df2['book_pages'].str.replace(' pages', '').astype('int')
book_df2

<ipython-input-6-0fc9ceff56a1>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  book_df2['book_pages_count'] = book_df2['book_pages'].str.replace(' pages', '').astype('int')


book_authors                                          book_desc  \
0       Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
1      Madeline Miller  In the house of Helios, god of the sun and mig...   
2          Holly Black  Of course I want to be like them. They’re beau...   
3            A.J. Finn  Anna Fox lives alone—a recluse in her New York...   
4        Tara Westover  An unforgettable memoir in the tradition of Th...   
...                ...                                                ...   
1136    Alyssa Brandon  A teen werewolf finally meets her destined sou...   
1137  Crystal Hana Kim  An emotionally riveting debut novel about war,...   
1138     Leslie  Cohen  This is the classic tale of boy meets girl: Gi...   
1139  Sarah Beth Durst  The battle between vicious spirits and strong-...   
1140      Nisha Sharma  Winnie Mehta was never really convinced that R...   

     book_edition     book_format     book_isbn book_pages  book_rating  \
0             NaN  Kindle Edition           NaN  435 pages         4.33   
1             NaN       Hardcover  9.780000e+12  393 pages         4.34   
2             NaN       Hardcover  9.780000e+12  370 pages         4.18   
3             NaN  Kindle Edition           NaN  449 pages         3.97   
4             NaN       Hardcover  9.780000e+12  334 pages         4.47   
...           ...             ...           ...        ...          ...   
1136          NaN       Hardcover  9.780000e+12  304 pages         3.33   
1137          NaN       Hardcover  9.780000e+12  417 pages         3.69   
1138          NaN  Kindle Edition           NaN  320 pages         3.40   
1139          NaN       Hardcover  9.780000e+12  419 pages         3.86   
1140          NaN       Hardcover  9.780000e+12  304 pages         3.64   

      book_rating_count  book_review_count  \
0                146505              17438   
1                 61357               9369   
2                 62897              11948   
3                175678              20645   
4                128337              16424   
...                 ...                ...   
1136                359                 88   
1137               1255                259   
1138               1300                272   
1139                907                205   
1140               1322                395   

                              book_title  \
0                        The Great Alone   
1                                  Circe   
2                       The Cruel Prince   
3                The Woman in the Window   
4                               Educated   
...                                  ...   
1136                        Bound to You   
1137                     If You Leave Me   
1138  This Love Story Will Self-Destruct   
1139                 The Queen of Sorrow   
1140         My So-Called Bollywood Life   

                                                 genres  \
0     Fiction|Historical|Historical Fiction|Audioboo...   
1     Fantasy|Fiction|Fantasy|Mythology|Historical|H...   
2                       Fantasy|Young Adult|Fairies|Fae   
3     Mystery|Thriller|Fiction|Suspense|Thriller|Mys...   
4             Nonfiction|Autobiography|Memoir|Biography   
...                                                 ...   
1136  Fantasy|Paranormal|Romance|Fantasy|Shapeshifte...   
1137  Historical|Historical Fiction|Fiction|Historic...   
1138  Romance|Fiction|Womens Fiction|Chick Lit|Conte...   
1139                          Fantasy|Young Adult|Adult   
1140  Young Adult|Contemporary|Romance|Fiction|Young...   

                                              image_url  book_pages_count  
0     https://images.gr-assets.com/books/1501852423l...               435  
1     https://images.gr-assets.com/books/1508879575l...               393  
2     https://images.gr-assets.com/books/1493047153l...               370  
3     https://images.gr-assets.com/books/1528225499l...               449  
4     https:/

Q5: What is the median number of pages?

In [7]:
book_df2['book_pages_count'].median()

326.5

Q6: What are the minimum and maximum numbers of pages?

In [8]:
book_df2['book_pages_count'].min()

0

In [9]:
book_df2['book_pages_count'].max()

992

Q7: Does having more than 1 author result in a longer book on average? What is the average number of pages for books written by a single author? What is it when there two authors? How about three authors? (We will do more careful analysis for these types of questions later. For now, we just want you to practice using some DataFrame functionalities).

In [10]:
book_df2[book_df2['book_authors'].str.contains('\|')]['book_pages_count'].mean()

280.0

In [11]:
book_df2[~book_df2['book_authors'].str.contains('\|')]['book_pages_count'].mean()

333.3510292524377

Having more than 1 author result in a shorter book on average.
The average number of pages for books written by a single author is 334.2275641025641 pages.

In [12]:
book_df2[book_df2['book_authors'].str.count('\|') == 1]['book_pages_count'].mean()

274.77777777777777

The average number of pages for books when there two authors is 280.26732673267327 pages.

In [13]:
book_df2[book_df2['book_authors'].str.count('\|') == 2]['book_pages_count'].mean()

277.75

The average number of pages for books when there two authors is 277.75 pages.

## Rating

Q8: How many books have at least a rating of 4.5?

In [14]:
book_df[book_df['book_rating'] >= 4.5]

book_authors  \
9                                       Neal Shusterman   
14                                          Delia Owens   
19                            Amie Kaufman|Jay Kristoff   
43                                       Pepper Winters   
45                                        Sarah J. Maas   
...                                                 ...   
955                      Brian K. Vaughan|Fiona Staples   
1002                                       Kester Grant   
1016                                         Penny Reid   
1087  Glory Edim|Jesmyn Ward|Lynn Nottage|Jacqueline...   
1122                                     Colleen Hoover   

                                              book_desc book_edition  \
9     Rowan has gone rogue, and has taken it upon hi...          NaN   
14    For years, rumors of the “Marsh Girl” have hau...          NaN   
19    Kady, Ezra, Hanna, and Nik narrowly escaped wi...          NaN   
43    “What do you do when you meet your soul mate? ...          NaN   
45    Years in the making, Sarah J. Maas’s #1 New Yo...          NaN   
...                                                 ...          ...   
955   The multiple Eisner Award-winning series retur...          NaN   
1002  A diverse fantasy reimagining of Les Misérable...          NaN   
1016  25 illustrations and doodles from the Knitting...  1st Edition   
1087  An inspiring collection of essays by black wom...          NaN   
1122  Lowen Ashleigh is a struggling writer on the b...          NaN   

         book_format     book_isbn book_pages  book_rating  book_rating_count  \
9          Hardcover  9.780000e+12  504 pages         4.53              22823   
14             ebook           NaN  384 pages         4.52              40414   
19         Hardcover  9.780000e+12  615 pages         4.58              18423   
43         Paperback  9.780000e+12  386 pages         4.50               5008   
45         Hardcover  9.780000e+12  992 pages         4.67              41705   
...              ...           ...        ...          ...                ...   
955   Kindle Edition           NaN  160 pages         4.57               7682   
1002             NaN           NaN  464 pages         4.60                 20   
1016       Paperback  9.780000e+12   66 pages         4.85                 26   
1087       Hardcover  9.780000e+12  272 pages         4.50                360   
1122  Kindle Edition           NaN  250 pages         4.56               9205   

      book_review_count                                         book_title  \
9                  4620                                        Thunderhead   
14                 5291                            Where the Crawdads Sing   
19                 4809                                            Obsidio   
43                 1021                             The Boy and His Ribbon   
45                 7227                                     Kingdom of Ash   
...                 ...                                                ...   
955                1779                                       Saga, Vol. 9   
1002                 25                                A Court of Miracles   
1016                  4                 Knitting in the City Coloring Book   
1087                115  Well-Read Black Girl: Finding Our Stories, Dis...   
1122               2493                                             Verity   

                                                 genres  \
9     Young Adult|Science Fiction|Science Fiction|Dy...   
14    Fiction|Historical|Historical Fiction|Mystery|...   
19                          Science Fiction|Young Adult   
43                  Romance|New Adult|Dark|Contemporary   
45                Fantasy|Young Adult|New Adult|Romance   
...                                                 ...   
955   Sequential Art|Graphic Novels|Sequential Art|C...   
1002  Fantasy|Young Adult|Historical|Historical Fict...   
1016                        

65 books have at least a rating of 4.5.

Q9: How about at most a rating of 4?

In [15]:
book_df[book_df['book_rating'] <= 4]

book_authors  \
3                          A.J. Finn   
8     Greer Hendricks|Sarah Pekkanen   
10                     Sarah J. Maas   
11                   Chloe  Benjamin   
13           Helen Hoang|Helen Hoang   
...                              ...   
1136                  Alyssa Brandon   
1137                Crystal Hana Kim   
1138                   Leslie  Cohen   
1139                Sarah Beth Durst   
1140                    Nisha Sharma   

                                              book_desc    book_edition  \
3     Anna Fox lives alone—a recluse in her New York...             NaN   
8     When you read this book, you will make many as...             NaN   
10    Hope warms the coldest night.Feyre, Rhys, and ...             NaN   
11    If you knew the date of your death, how would ...             NaN   
13    A heartwarming and refreshing debut novel that...  1st US edition   
...                                                 ...             ...   
1136  A teen werewolf finally meets her destined sou...             NaN   
1137  An emotionally riveting debut novel about war,...             NaN   
1138  This is the classic tale of boy meets girl: Gi...             NaN   
1139  The battle between vicious spirits and strong-...             NaN   
1140  Winnie Mehta was never really convinced that R...             NaN   

         book_format     book_isbn book_pages  book_rating  book_rating_count  \
3     Kindle Edition           NaN  449 pages         3.97             175678   
8          Hardcover  9.780000e+12  346 pages         3.85              89835   
10         Hardcover  9.780000e+12  229 pages         3.92              64962   
11         Hardcover  9.780000e+12  346 pages         3.73              61815   
13    Kindle Edition           NaN  336 pages         4.00              33760   
...              ...           ...        ...          ...                ...   
1136       Hardcover  9.780000e+12  304 pages         3.33                359   
1137       Hardcover  9.780000e+12  417 pages         3.69               1255   
1138  Kindle Edition           NaN  320 pages         3.40               1300   
1139       Hardcover  9.780000e+12  419 pages         3.86                907   
1140       Hardcover  9.780000e+12  304 pages         3.64               1322   

      book_review_count                          book_title  \
3                 20645             The Woman in the Window   
8                 11331                 The Wife Between Us   
10                 9583      A Court of Frost and Starlight   
11                 7510                    The Immortalists   
13                 6502                   The Kiss Quotient   
...                 ...                                 ...   
1136                 88                        Bound to You   
1137                259                     If You Leave Me   
1138                272  This Love Story Will Self-Destruct   
1139                205                 The Queen of Sorrow   
1140                395         My So-Called Bollywood Life   

                                                 genres  \
3     Mystery|Thriller|Fiction|Suspense|Thriller|Mys...   
8     Thriller|Fiction|Mystery|Suspense|Thriller|Mys...   
10                Fantasy|New Adult|Romance|Young Adult   
11    Fiction|Fantasy|Contemporary|Historical|Histor...   
13    Romance|Contemporary|Fiction|Adult|Romance|Con...   
...                                                 ...   
1136  Fantasy|Paranormal|Romance|Fantasy|Shapeshifte...   
1137  Historical|Historical Fiction|Fiction|Historic...   
1138  Romance|Fiction|Womens Fiction|Chick Lit|Conte...   
1139                          Fantasy|Young Adult|Adult   
1140  Young Adult|Contemporary|Romance|Fiction|Young...   

                                              image_url  
3     https://images.gr-assets.com/books/1528225499l...  
8     https://images.gr-assets.com/books/1496862122l...  
10    https://images.gr-asset

489 books have at most a rating of 4.

Q10: Discretize (i.e. round down) the ratings. The resulting ratings should have one of the following values: 1,2,3,4 or 5.

In [16]:
book_df['book_rating_rounded'] = book_df['book_rating'].astype('int')
book_df

<ipython-input-16-5134f2a61c3b>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  book_df['book_rating_rounded'] = book_df['book_rating'].astype('int')


book_authors                                          book_desc  \
0       Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
1      Madeline Miller  In the house of Helios, god of the sun and mig...   
2          Holly Black  Of course I want to be like them. They’re beau...   
3            A.J. Finn  Anna Fox lives alone—a recluse in her New York...   
4        Tara Westover  An unforgettable memoir in the tradition of Th...   
...                ...                                                ...   
1136    Alyssa Brandon  A teen werewolf finally meets her destined sou...   
1137  Crystal Hana Kim  An emotionally riveting debut novel about war,...   
1138     Leslie  Cohen  This is the classic tale of boy meets girl: Gi...   
1139  Sarah Beth Durst  The battle between vicious spirits and strong-...   
1140      Nisha Sharma  Winnie Mehta was never really convinced that R...   

     book_edition     book_format     book_isbn book_pages  book_rating  \
0             NaN  Kindle Edition           NaN  435 pages         4.33   
1             NaN       Hardcover  9.780000e+12  393 pages         4.34   
2             NaN       Hardcover  9.780000e+12  370 pages         4.18   
3             NaN  Kindle Edition           NaN  449 pages         3.97   
4             NaN       Hardcover  9.780000e+12  334 pages         4.47   
...           ...             ...           ...        ...          ...   
1136          NaN       Hardcover  9.780000e+12  304 pages         3.33   
1137          NaN       Hardcover  9.780000e+12  417 pages         3.69   
1138          NaN  Kindle Edition           NaN  320 pages         3.40   
1139          NaN       Hardcover  9.780000e+12  419 pages         3.86   
1140          NaN       Hardcover  9.780000e+12  304 pages         3.64   

      book_rating_count  book_review_count  \
0                146505              17438   
1                 61357               9369   
2                 62897              11948   
3                175678              20645   
4                128337              16424   
...                 ...                ...   
1136                359                 88   
1137               1255                259   
1138               1300                272   
1139                907                205   
1140               1322                395   

                              book_title  \
0                        The Great Alone   
1                                  Circe   
2                       The Cruel Prince   
3                The Woman in the Window   
4                               Educated   
...                                  ...   
1136                        Bound to You   
1137                     If You Leave Me   
1138  This Love Story Will Self-Destruct   
1139                 The Queen of Sorrow   
1140         My So-Called Bollywood Life   

                                                 genres  \
0     Fiction|Historical|Historical Fiction|Audioboo...   
1     Fantasy|Fiction|Fantasy|Mythology|Historical|H...   
2                       Fantasy|Young Adult|Fairies|Fae   
3     Mystery|Thriller|Fiction|Suspense|Thriller|Mys...   
4             Nonfiction|Autobiography|Memoir|Biography   
...                                                 ...   
1136  Fantasy|Paranormal|Romance|Fantasy|Shapeshifte...   
1137  Historical|Historical Fiction|Fiction|Historic...   
1138  Romance|Fiction|Womens Fiction|Chick Lit|Conte...   
1139                          Fantasy|Young Adult|Adult   
1140  Young Adult|Contemporary|Romance|Fiction|Young...   

                                              image_url  book_rating_rounded  
0     https://images.gr-assets.com/books/1501852423l...                    4  
1     https://images.gr-assets.com/books/1508879575l...                    4  
2     https://images.gr-assets.com/books/1493047153l...                    4  
3     https://images.gr-assets.com/books/1528225499l...                    3 

Q11: For each of the discretized ratings (1,2,3,4,5), what is the average number 
of authors?

In [17]:
book_df['book_authors_count'] = book_df['book_authors'].str.count('\|') + 1
book_df.groupby(['book_rating_rounded'])['book_authors_count'].mean()

<ipython-input-17-cac4caaa3b23>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  book_df['book_authors_count'] = book_df['book_authors'].str.count('\|') + 1


book_rating_rounded
2    1.333333
3    1.222458
4    1.315133
5    1.000000
Name: book_authors_count, dtype: float64

Q12: How about the average number of reviews?

In [18]:
book_df.groupby(['book_rating_rounded'])['book_review_count'].mean()

book_rating_rounded
2    251.333333
3    884.114407
4    859.383775
5      0.777778
Name: book_review_count, dtype: float64

## Genres

Q13: Create a new DataFrame, exploding the rows with multiple genres such that it is one row per genre/book.

In [19]:
book_df3=book_df.drop('genres', axis=1).join(book_df['genres'].str.split('|', expand=True).stack().reset_index(level=1, drop=True).rename('genres'))
book_df3

book_authors                                          book_desc  \
0     Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
0     Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
0     Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
0     Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
0     Kristin Hannah  Alaska, 1974.Unpredictable. Unforgiving. Untam...   
...              ...                                                ...   
1140    Nisha Sharma  Winnie Mehta was never really convinced that R...   
1140    Nisha Sharma  Winnie Mehta was never really convinced that R...   
1140    Nisha Sharma  Winnie Mehta was never really convinced that R...   
1140    Nisha Sharma  Winnie Mehta was never really convinced that R...   
1140    Nisha Sharma  Winnie Mehta was never really convinced that R...   

     book_edition     book_format     book_isbn book_pages  book_rating  \
0             NaN  Kindle Edition           NaN  435 pages         4.33   
0             NaN  Kindle Edition           NaN  435 pages         4.33   
0             NaN  Kindle Edition           NaN  435 pages         4.33   
0             NaN  Kindle Edition           NaN  435 pages         4.33   
0             NaN  Kindle Edition           NaN  435 pages         4.33   
...           ...             ...           ...        ...          ...   
1140          NaN       Hardcover  9.780000e+12  304 pages         3.64   
1140          NaN       Hardcover  9.780000e+12  304 pages         3.64   
1140          NaN       Hardcover  9.780000e+12  304 pages         3.64   
1140          NaN       Hardcover  9.780000e+12  304 pages         3.64   
1140          NaN       Hardcover  9.780000e+12  304 pages         3.64   

      book_rating_count  book_review_count                   book_title  \
0                146505              17438              The Great Alone   
0                146505              17438              The Great Alone   
0                146505              17438              The Great Alone   
0                146505              17438              The Great Alone   
0                146505              17438              The Great Alone   
...                 ...                ...                          ...   
1140               1322                395  My So-Called Bollywood Life   
1140               1322                395  My So-Called Bollywood Life   
1140               1322                395  My So-Called Bollywood Life   
1140               1322                395  My So-Called Bollywood Life   
1140               1322                395  My So-Called Bollywood Life   

                                              image_url  book_rating_rounded  \
0     https://images.gr-assets.com/books/1501852423l...                    4   
0     https://images.gr-assets.com/books/1501852423l...                    4   
0     https://images.gr-assets.com/books/1501852423l...                    4   
0     https://images.gr-assets.com/books/1501852423l...                    4   
0     https://images.gr-assets.com/books/1501852423l...                    4   
...                                                 ...                  ...   
1140  https://images.gr-assets.com/books/1510586092l...                    3   
1140  https://images.gr-assets.com/books/1510586092l...                    3   
1140  https://images.gr-assets.com/books/1510586092l...                    3   
1140  https://images.gr-assets.com/books/1510586092l...                    3   
1140  https://images.gr-assets.com/books/1510586092l...                    3   

      book_authors_count                    genres  
0                      1                   Fiction  
0                      1                Historical  
0                      1        Historical Fiction  
0                      1                 Audiobook  
0                      1                Historical  
...                  ...  

Q14: What is the average rating of books from different genres? What is the median?

In [20]:
book_df3.groupby(['genres'])['book_rating'].mean()

genres
16th Century                      4.130000
20th Century                      4.000000
Abandoned                         3.557692
Abuse                             4.365000
Academic                          4.193333
                                    ...   
Young Adult                       3.932241
Young Adult Contemporary          3.770000
Young Adult Fantasy               3.923846
Young Adult Historical Fiction    3.900000
Young Adult Romance               3.680000
Name: book_rating, Length: 270, dtype: float64

In [21]:
book_df3.groupby(['genres'])['book_rating'].median()

genres
16th Century                      4.130
20th Century                      4.000
Abandoned                         3.630
Abuse                             4.300
Academic                          4.130
                                  ...  
Young Adult                       3.925
Young Adult Contemporary          3.785
Young Adult Fantasy               3.920
Young Adult Historical Fiction    3.900
Young Adult Romance               3.680
Name: book_rating, Length: 270, dtype: float64

Q15: What is the average number of authors of books from different genres? What is the median?

In [22]:
book_df3.groupby(['genres'])['book_authors_count'].mean()

genres
16th Century                      1.000000
20th Century                      1.000000
Abandoned                         1.000000
Abuse                             1.500000
Academic                          1.000000
                                    ...   
Young Adult                       1.405172
Young Adult Contemporary          1.125000
Young Adult Fantasy               1.076923
Young Adult Historical Fiction    1.000000
Young Adult Romance               1.000000
Name: book_authors_count, Length: 270, dtype: float64

In [23]:
book_df3.groupby(['genres'])['book_authors_count'].median()

genres
16th Century                      1.0
20th Century                      1.0
Abandoned                         1.0
Abuse                             1.5
Academic                          1.0
                                 ... 
Young Adult                       1.0
Young Adult Contemporary          1.0
Young Adult Fantasy               1.0
Young Adult Historical Fiction    1.0
Young Adult Romance               1.0
Name: book_authors_count, Length: 270, dtype: float64